File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res=0
        strs=""
        i=0
        while i<len(s):
            if s[i] not in strs:
                strs+=(s[i])
                res= max(res,len(strs))
                i+=1
                
            else:
                ind= strs.index(s[i])
                strs=strs[ind+1:]+s[i]
                i+=1
                

            
        return res
                

File: Data Structures & Algorithms/partition-labels/submission-3.java
class Solution {
    public List<Integer> partitionLabels(String s) {
        HashMap<Character, Integer> idx= new HashMap<>();



        List<Integer> res=  new ArrayList<>();
        

        for(int j = 0; j< s.length(); j++){
            
            idx.put(s.charAt(j), j);
        }

        int lastIndex = 0;
        int end = 0;

        


        for(int i =0 ; i<s.length(); i++){
            
            
            end = Math.max(end, idx.get(s.charAt(i)));
            if(i == end){
                res.add(end - lastIndex + 1);
                lastIndex = end + 1;
            }
            

        }

        return res;
        
    }
}
